### ISSUE TITLE
InfluxDB Flux: cannot access calculated and mapped values

### ISSUE THREAD

**Issue by LaXiS96** [grafana/grafana#28078]
**What happened**:
Consider this Flux query:
```
from(bucket: "DataCollector")
  |> range(start: -30m)
  |> filter(fn: (r) =>
    r._measurement == "OpcUa" and
    r._field == "Alarm")
  |> difference()
  |> elapsed()
  |> map(fn: (r) => ({
    r with state:
      if r._value == 1 then "0"
      else if r._value == -1 then "1"
      else ""
    }))
  |> group(columns: ["LineDescription", "LayoutId"])
  |> yield()
```
`OpcUa` is a measurement where `LineDescription` and `LayoutId` are the only tags and `Alarm` is one of many fields.
The `Alarm` field value is 0 or 1, representing the new state after the change that happened at that timestamp. See the raw field data screenshot: https://pasteboard.co/Juws1ho.png

When using this query as source for a table, all that is returned is a single column (grouped by LineDescription, LayoutId and the mapped `state` column) and its value is the result of the difference() function (since its result is put into the _value column).

Grafana screenshot: https://pasteboard.co/JuwcuFS.png
Chronograf screenshot: https://pasteboard.co/JuweoUg.png
As you can see:
- `state` is grouped with `LineDescription` and `LayoutId`
- the elapsed() value is nowhere to be found
- all grouping combobox values read as `OpcUa`

**What you expected to happen**:
- `state` should be its own column, not part of a group header (unless explicitly indicated)
- `elapsed` should be its own column
- the grouping combobox should show one `LineDescription=xxx,LayoutId=yyy` for each group, like Chronograf does

**How to reproduce it (as minimally and precisely as possible)**:
Possibly, by using any query that uses map() and/or elapsed() functions. There may also be other functions which return calculated columns, where this problem presents itself.

**Anything else we need to know?**:
I'm no InfluxDB nor Grafana expert, so I may be misunderstanding how Flux returns data, so maybe this is not even a bug.

**Environment**:
- Grafana version: Docker image grafana/grafana:7.2.0
- Data source type & version: InfluxDB v1.8.3 with Flux enabled
- OS Grafana is installed on: RancherOS
- User OS & Browser: Windows 10 18362.1082 and Chrome 85
- Grafana plugins: none
- Others: none

*EDIT:* added missing InfluxDB version, minor precisation in reproduction section

**Comment by aknuds1** [comment]
@ryantxu Could you help out on this one?

**Comment by marefr** [comment]
I may be wrong, but I think you have to use a transformation to achieve what you want. I have this similar example with 3 series with each having label key server and values A, B and C:
![image](https://user-images.githubusercontent.com/1668778/95392697-7fdcfc80-08f9-11eb-96b6-c3edd61f410b.png)i

Adding a labels to fields transformation I think the result is in line with what you want:
![image](https://user-images.githubusercontent.com/1668778/95392784-abf87d80-08f9-11eb-9be0-8be321c76c02.png)

**Comment by ryantxu** [comment]
@LaXiS96 -- are you able to share the CSV output from chronograf?

I am not totally fluent with flux, but i think you can force it into a table mode rather than the timeseries you are currently returning.  as @marefr points out, you can always convert the labels back to columns -- they are just constants :)

**Comment by LaXiS96** [comment]
@ryantxu Here are the Chronograf csv exports for the same time period (2020-10-07 13:00 to 14:00 UTC):
[raw data](https://github.com/grafana/grafana/files/5345816/raw.txt)
[query results](https://github.com/grafana/grafana/files/5345815/query.txt)

@marefr's suggestion looked like a promising workaround until I noticed that the `state` column is always showing one unique value per group, which is wrong. It should be alternating the same way the `Alarm` column does. You can see that in the query results csv, the values are as I would expect them to be, while Grafana shows something different.
See this screenshot: https://pasteboard.co/JuEKbQm.png.

Thank you

**Comment by LaXiS96** [comment]
I think I found a way around, but I don't understand if this is the "correct" way to extract data via Flux and if Grafana could do something to make this more intuitive.

Adding this as a last function in the pipeline:
```
  |> pivot(rowKey:["_time", "state", "elapsed"], columnKey: ["_field"], valueColumn: "_value")
```
returns a single record for each unique combination of `_time`, `state` and `elapsed` values, with the side effect that those columns are now visible in Grafana.

![image](https://user-images.githubusercontent.com/8070973/95449622-52c63380-0965-11eb-980f-bca416c88bd6.png)

Still, the grouping combobox is broken...

**Comment by R-Studio** [comment]
I think I have the same issue with InfluxDB with Flux as the query language. 
But for the InfluxDB datasource with InfluxQL as the query language there is an option "**FORMAT AS Table**" that solves the problem.
![image](https://user-images.githubusercontent.com/7498854/100746747-692ac180-33e1-11eb-8424-d606179ef329.png)

**-> The best solution is maybe to add this option "FORMAT AS Table" also for the Flux datasource.**

**Comment by R-Studio** [comment]
@ryantxu & @LaXiS96 any news on this?

**Comment by LaXiS96** [comment]
@R-Studio Sorry, I haven't had time to investigate deeper.

I don't know if the problem is with the way I'm extracting data (since adding a `pivot()` "fixed" the problem) or with Grafana's visualization of the original data.
The main problem I see is that Flux does not follow the usual "table" pattern as SQL relational databases do, there is no concept of "record" so each field and value combo is a "record" by itself and can be/is returned as such.

**Comment by R-Studio** [comment]
@LaXiS96  unfortunately in some UseCase the workaround with `pivot()` works, but **in a many cases not**.
For example I will display all data from a measurement as a table but grafana groups per series:
![image](https://user-images.githubusercontent.com/7498854/104452366-39a3d300-55a3-11eb-909a-2f5cf2f30eb4.png)
Also if I add the Grafana transform function "Series to rows", it doesn't work. 

_In Grafana 7.x there are a lot of "bugs" or workarounds in context with the datasource Flux (InfluxDB). That is very sad._

**Comment by PaulSchulz** [comment]
I suspect that I am hitting a similar problem when trying to extract lat/long data for the Grafana WorldMap.

I have a flux query which gives me the table data I require on the influx command line, but does not work inside Gafana (which appears to get stuck on the '_value' column somehow.) (NB: I am only able to push long/lat into influx as strings, hence the conversion step towards the end.)

from(bucket: "iot")
|> range(start: -24h)
|> filter(fn: (r) => r["_measurement"] == "gateway")
|> keep(columns: ["_time", "name", "longitude", "latitude", "_field", "_value"])
|> group(columns: ["name"])
|> sort(columns: ["_time"], desc: true)
|> first()
|> group(columns: ["_field"])
|> drop(columns: ["_field","_time"])
|> map(fn:(r) => ({ r with longitude: float(v: r.longitude) }))
|> map(fn:(r) => ({ r with latitude: float(v: r.latitude) }))
|> map(fn:(r) => ({ metric: r._value, latitude: r.latitude, longitude: r.longitude, name: r.name }))
|> drop(columns: ["result","table"])

**Comment by gabor** [comment]
@PaulSchulz @LaXiS96 @R-Studio the way grafana currently parses the result from influxdb is that if we find a "_value" column we assume time-series and we kind of ignore other value-columns. we have plans to improve this ( https://github.com/grafana/grafana/issues/30085 ) , a quick workaround is to simply make sure there is no "_value" column in the returned data, for example by adding
```
|> rename(columns: {_value: "something"})
```

**Comment by gabor** [comment]
(i closed this by mistake, reopening)

**Comment by R-Studio** [comment]
> @PaulSchulz @LaXiS96 @R-Studio the way grafana currently parses the result from influxdb is that if we find a "_value" column we assume time-series and we kind of ignore other value-columns. we have plans to improve this ( #30085 ) , a quick workaround is to simply make sure there is no "_value" column in the returned data, for example by adding
> 
> ```
> |> rename(columns: {_value: "something"})
> ```
> 
> at the end of the query. does this workaround help?

Thanks this for your help but this workarounds does not always helps (for example in the pie chart panel). Because the renaming is a hardcoded string and you can't use another field or something.

**Comment by gabor** [comment]
hmm.. interesting.. could you give more info about the pie-chart problem? (we have some plans about a potential improvement to this problem here, i just want to make sure we cover as many problem-cases as possible)

**Comment by R-Studio** [comment]
@gabor 

**My Query**
```
from(bucket: bucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => 
    r._measurement == "asset_hyperv_local" and
    r._field == "physicalmemory"
  )
  |> last()
  |> group(columns: ["sla"])
  |> count()
```

**InfluxDB Explore**
![image](https://user-images.githubusercontent.com/7498854/114135682-e58e5d80-9909-11eb-8b21-254efe951def.png)

**Grafana**
![image](https://user-images.githubusercontent.com/7498854/114135788-0bb3fd80-990a-11eb-80c4-0de7d671c72c.png)

**What I want**
I want the number and percentage of each "sla", like in the InfluxDB Explorer but displayed as a pie chart.

**What I expect**
I expect that Grafana displays the label the value of the label "sla" instead of "_value".
_When I use the function "rename()" then I am not able to rename based on the label "sla" (I think this is because rename uses a hardcoded string and cannot be used dynamically with an label)._

**Comment by gabor** [comment]
when https://github.com/grafana/grafana/issues/32842 happens it should help with this situation (i mean the general case here, did not look into the "pie chart issue" yet)

**Comment by gabor** [comment]
@R-Studio (regarding the pie chart issue) could you get the CSV data from the flux-query too? (there is a download-csv button in chronograf, next to the view-raw-data toggle)

**Comment by R-Studio** [comment]
@gabor thanks for your reply and of course here is the CSV:
```
#group,false,false,true,false
#datatype,string,long,string,long
#default,_result,,,
,result,table,sla,_value
,,0,-,7
,,1,Bronze,26
,,2,Gold,82
,,3,Platin,63
,,4,PreProd,2
,,5,Rhodium,16
,,6,Silber,90
,,7,Silber/Gold,190
```

**Comment by pheetr** [comment]
> 
> 
> @PaulSchulz @LaXiS96 @R-Studio the way grafana currently parses the result from influxdb is that if we find a "_value" column we assume time-series and we kind of ignore other value-columns. we have plans to improve this ( #30085 ) , a quick workaround is to simply make sure there is no "_value" column in the returned data, for example by adding
> 
> ```
> |> rename(columns: {_value: "something"})
> ```
> 
> at the end of the query. does this workaround help?

Hi, when I try this, Grafana doesn't visualize the values in the graph for some reason. Is there a solution to this, some option I may not be aware of (I'm fairly new to Grafana)? Thx.

![image](https://user-images.githubusercontent.com/44616229/116685324-0d715e00-a9aa-11eb-96b6-2a925ce8d91c.png)

**Comment by gabor** [comment]
@pheetr could you get me 2 things?
1. the query you used
2. the CSV-format response from influxdb. the simplest way to get it is to use the default influxdb gui (chronograf), explore-mode, paste the query in it, execute it and there is a download-csv button.

of course, sensitive parts removed.

**Comment by pheetr** [comment]
Sure, @gabor, please see below. Thx.
```
from(bucket: "test")
|> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "win_process")
  |> filter(fn: (r) => r["_field"] == "Private_Bytes")
  |> filter(fn: (r) => r["instance"] =~ /firefox.*/)
  |> group() // resets group key putting all tables in one group
  |> aggregateWindow(every: 30s, fn: sum)
  |> drop(columns: ["host"])
  |> rename(columns: {_value: "FF"})
```

Response (GitHub threw an error when trying to attach the csv): https://pastebin.com/DeSQncEJ

**Comment by gabor** [comment]
@pheetr thanks for the data, i will look into this, i created a separate issue for your problem ( https://github.com/grafana/grafana/issues/33675 ), because it might or might not be the same as this issue, and this one is getting very long :-) please do subscribe to that issue so you get notifications, and we can continue the discussion there.

**Comment by gabor** [comment]
we changed how complex flux queries are parsed. it is available in 8.0.0-beta1 or later. it should fix problems like this issue probably ("tag" columns and "data" columns are parsed more correctly now and no data-column is ignored), but the response format might be slightly different because of this.

if it does not solve your issue completely, please open a new issue (i'd like to not mix old-behavior and new-behavior in the same issue, to keep issues easy to understand).
